File: 2023/Feb/Feb-6/LinkedListMiddle/Solution.java
// Node of a linked list
 class Node {
   int data;
    Node next;
    Node(int d)  { data = d;  next = null; }    
}


public class Solution
{
    public static void main(String[] args) {
        
        Node head=  new Node(1);
        Node tail = head;
        for (int i = 2; i < 6; i++) {
            Node n = new Node(i);
            if(head.equals(tail) ){
                tail = new Node(i);
                head.next = tail;
            }
            else{
                tail.next= n;
                tail = n;
            }
                
        }

        int ans=  Solution.getMiddle(head);
        System.out.println("Ans: "+ ans);

    }

    public static int getMiddle(Node head)
    {
        int count =0, ans =-1;
        
        Node itr = head;
        
        while(itr.next !=  null){
            itr = itr.next;
            count++;
        }
        System.out.println("num of elements: "+count);
        itr = head;
        for(int i =0; i< count/2 ; i++ )
        {
            
            itr =itr.next;            
        }
        ans = itr.data;
        
        
        return ans;
    }

}

File: 2023/Feb/Feb-6/LinkedListMiddle/Solution.cpp
/* Link list Node
 */
#include <bits/stdc++.h>
struct Node
{
    int data;
    Node *next;

    Node(int x)
    {
        data = x;
        next = NULL;
    }
};

class Solution
{
public:
    /* Should return data of middle node. If linked list is empty, then  -1*/
    int getMiddle(Node *head)
    {
        Node *fast = head, *slow = head;
        if(!head) return -1;
        bool even = false;
        while (fast->next != nullptr && fast->next->next != nullptr)
        {
            fast = fast->next->next;
            if(! fast) even = true;
            slow = slow->next;
        }
        
        int ans ;
        if(!even)

        ans = slow->next->data;
        else 
        ans = slow->data;

        return ans;
        
    }

};